File: mifosng-android/src/main/java/com/mifos/mifosxdroid/online/savingsaccountapproval/SavingsAccountApprovalPresenter.java
package com.mifos.mifosxdroid.online.savingsaccountapproval;

import com.mifos.api.GenericResponse;
import com.mifos.api.datamanager.DataManagerSavings;
import com.mifos.mifosxdroid.base.BasePresenter;
import com.mifos.objects.accounts.loan.SavingsApproval;
import com.mifos.utils.MFErrorParser;

import javax.inject.Inject;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.observers.DisposableObserver;
import io.reactivex.schedulers.Schedulers;


/**
 * Created by Rajan Maurya on 09/06/16.
 */
public class SavingsAccountApprovalPresenter extends BasePresenter<SavingsAccountApprovalMvpView> {

    private final DataManagerSavings dataManagerSavings;
    private DisposableObserver<GenericResponse> disposableObserver;

    @Inject
    public SavingsAccountApprovalPresenter(DataManagerSavings dataManager) {
        dataManagerSavings = dataManager;
    }

    @Override
    public void attachView(SavingsAccountApprovalMvpView mvpView) {
        super.attachView(mvpView);
    }

    @Override
    public void detachView() {
        super.detachView();
        if (disposableObserver != null) disposableObserver.dispose();
    }

    public void approveSavingsApplication(int savingsAccountId, SavingsApproval savingsApproval) {
        checkViewAttached();
        getMvpView().showProgressbar(true);
        if (disposableObserver != null) disposableObserver.dispose();
        disposableObserver = dataManagerSavings
                .approveSavingsApplication(savingsAccountId, savingsApproval)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribeWith(new DisposableObserver<GenericResponse>() {

                    @Override
                    public void onError(Throwable e) {
                        getMvpView().showProgressbar(false);
                        getMvpView().showError(MFErrorParser.errorMessage(e));
                    }

                    @Override
                    public void onComplete() {

                    }

                    @Override
                    public void onNext(GenericResponse genericResponse) {
                        getMvpView().showProgressbar(false);
                        getMvpView().showSavingAccountApprovedSuccessfully(genericResponse);
                    }
                });
    }
}

File: mifosng-android/src/main/java/com/mifos/mifosxdroid/online/savingsaccountapproval/SavingsAccountApprovalPresenter.kt
package com.mifos.mifosxdroid.online.savingsaccountapproval

import com.mifos.api.GenericResponse
import com.mifos.api.datamanager.DataManagerSavings
import com.mifos.mifosxdroid.base.BasePresenter
import com.mifos.objects.accounts.loan.SavingsApproval
import com.mifos.utils.MFErrorParser
import rx.Subscriber
import rx.Subscription
import rx.android.schedulers.AndroidSchedulers
import rx.schedulers.Schedulers
import javax.inject.Inject

/**
 * Created by Rajan Maurya on 09/06/16.
 */
class SavingsAccountApprovalPresenter @Inject constructor(private val dataManagerSavings: DataManagerSavings) : BasePresenter<SavingsAccountApprovalMvpView?>() {
    private var mSubscription: Subscription? = null
    override fun attachView(mvpView: SavingsAccountApprovalMvpView?) {
        super.attachView(mvpView)
    }

    override fun detachView() {
        super.detachView()
        if (mSubscription != null) mSubscription!!.unsubscribe()
    }

    fun approveSavingsApplication(savingsAccountId: Int, savingsApproval: SavingsApproval?) {
        checkViewAttached()
        mvpView!!.showProgressbar(true)
        if (mSubscription != null) mSubscription!!.unsubscribe()
        mSubscription = dataManagerSavings
                .approveSavingsApplication(savingsAccountId, savingsApproval)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribe(object : Subscriber<GenericResponse?>() {
                    override fun onCompleted() {}
                    override fun onError(e: Throwable) {
                        mvpView!!.showProgressbar(false)
                        mvpView!!.showError(MFErrorParser.errorMessage(e))
                    }

                    override fun onNext(genericResponse: GenericResponse?) {
                        mvpView!!.showProgressbar(false)
                        mvpView!!.showSavingAccountApprovedSuccessfully(genericResponse)
                    }
                })
    }

}